### ISSUE TITLE
Modularize Ratatui crate

### ISSUE THREAD

**Issue by joshka** [ratatui/ratatui#1388]
## Problem
It would be nice to be able for widget crates to rely just on the widget / buffer / etc. types instead of the whole library.
Additionally by extracting a widget crate out of the library we might be able to have quicker incremental build times.
It would also be nice to be able to release non-breaking widget updates in point releases immediately rather than waiting for a library release.

## Solution
Decompose Ratatui into several crates:
- ratatui: the main core library, feature flags to include the parts
- ratatui-core: Widget, WidgetRef, StatefulWidget, StatefulWidgetRef, Terminal, Frame, Backend etc.
- ratatui-{crossterm,termion,termwiz}: backends
- ratatui-widgets: all the existing built-in widgets (currently this is something else, but I'd move the current library to a new name)

It's possible that we could also go more granular than this and have a crate per widget, but I think it's worth doing the first part and evaluating whether doing more is worthwhile.

This will be implemented by making the repo into a cargo workspace, moving the current source under /ratatui/src, and creating several more crates at the root. I know it may be tempting to not do a monorepo for this, but the administrative overhead of having a repo per crate seem prohibitive, while I've had reasonable success using mono-repos (e.g. tui-widgets crate).

The work to do this will be:
- make the actual refactoring changes
- make the CI / CD changes
- make the doc changes (fixup website links etc. to examples)
- test changes in various widget libraries

@ratatui/maintainers what do you think?

# Plan

- [x] move the folder, create the workspace #1459 
- [x] move the main types to ratatui-core #1460 
- [x] Move widgets into ratatui-widgets https://github.com/ratatui/ratatui/pull/1474
- [x] https://github.com/ratatui/ratatui/pull/1479
- [x] https://github.com/ratatui/ratatui/pull/1490
- [x] https://github.com/ratatui/ratatui/pull/1491
- [x] #1508 
- [x] Move examples into /examples - there's a good approach that tokio uses for this where each example gets a crate, and so can have their own dependencies listed etc rather than these being all smooshed into a single dev-dependencies lists.
- [x] Update / make sure that git cliff and release automation are working fine
- [ ] Make an alpha release and then test ratatui-core with some widget libraries
- [x] decide on whether to move frame / terminal to core, or leave them in Ratatui (I'm not certain about this right now, probably should go to core, but there's an argument against it too

**Comment by joshka** [comment]
# Blockers

I'm attempting a PoC of this, and documenting issues that come up

- crossterm backend code implements conversions to / from styles, colors, modifiers (attributes in crossterm). Because these items will be defined in separate crates, the orphan rule prevents these conversions from being defined in the backend crate. In theory these conversions should be seen as implementation details of the backend, but there may be code that relies on this. I'm not sure what the right thing to do with this is

**Comment by joshka** [comment]
PoC: https://github.com/ratatui/ratatui/pull/1389

**Comment by kdheepak** [comment]
I'm supportive of this! And in favor of a monorepo for this too! I've never done this myself though, so I'd have to learn more. I'm curious how ci/cd files will be managed because we wouldn't want to run ci in ratatui-core if changes were made in ratatui-widgets (or maybe it doesn't matter?). I'm sure there are some git cliff nuances here too.

**Comment by joshka** [comment]
Using release-plz, I just get release PRs for the specific crate that was changed in tui-widgets, so scoping down to specific folder changes is possible. Obv. we'd have a bit of a different approach.

Some pre-work for this is to remove all usage of the prelude. This will make the code diffs a bit smaller.

**Comment by orhun** [comment]
This sounds good in theory, but do we have any practical use cases for it right now? For example, are there any apps that have shown a need for this kind of split functionality? I’d like to understand why this might be necessary before proceeding with the implementation.

> Using release-plz, I just get release PRs for the specific crate that was changed in tui-widgets, so scoping down to specific folder changes is possible.

We probably need to fix the integration between git-cliff and release-plz in that case.

> Some pre-work for this is to remove all usage of the prelude. This will make the code diffs a bit smaller.

Left a comment in that PR.

**Comment by joshka** [comment]
Every widget library has the problem of having to update the ratatui version whenever we release a breaking change. Often we don't break the core, but a semver break for a widget is a semver break for ratatui. If instead a widget library relies on ratatui-core 0.1, then the widget library is compatible with all versions of ratatui that use ratatui-core 0.1. This makes updating apps to later versions of Ratatui easier.

This is a problem that happens already - https://github.com/joshka/tui-widgets/issues/31 is sort of related (@matta would have some perspective on this)

> We probably need to fix the integration between git-cliff and release-plz in that case.

The behavior I described is desired behavior, and it's how release-plz works. When I change only one crate, then I should see only a changelog entry for that one crate.

**Comment by orhun** [comment]
> Every widget library has the problem of having to update the ratatui version whenever we release a breaking change.

So in that case we can start by splitting `ratatui` and `ratatui-widgets`? Do we need to split the backends as well?

> The behavior I described is desired behavior, and it's how release-plz works. When I change only one crate, then I should see only a changelog entry for that one crate.

Yeah, but our current git-cliff configuration does not work with release-plz.

**Comment by joshka** [comment]
(I updated the comment to be more clear) I was suggesting that given that doing something different per folder is possible in an action, then we know that the approach of doing this is a possible one. I'm saying that this isn't a hard blocker, just one that requires some configuration changes to make work right.

> So in that case we can start by splitting ratatui and ratatui-widgets? Do we need to split the backends as well?

Yes, the first split would be widgets. To do that split requires Ratatui to depend on ratatui-widgets, and for ratatui-widgets to be able to import ratatui::widgets::Widget etc, so that needs to be moved into ratatui-core.

Yes, the backends should be split too. This isolates widget libraries from backend updates in the same way. E.g. crossterm 0.29 won't cause ratatui-core to have to bump, which means all widgets continue to work if there's an update in the backend.

**Comment by joshka** [comment]
# Crate Versioning

I'd suggest that we start the new crates at 0.1.0. This is because bumping core will likely happen irregularly, while bumping -widgets would happen much more often, so these should get out of sync any time we semver bump one but not the other.

The other thing we could do is start them at whatever the latest Ratatui version is at the time of the split (0.29 / 0.30). This may be useful in the short term to help show that these have a lineage back to the current Ratatui version, but this becomes less useful in the longer term.

I'm 60/40 for using 0.1.0 rather than 0.29.0 for these, but happy to be convinced of the alternative approach.

**Comment by kdheepak** [comment]
> ratatui: the main core library, feature flags to include the parts

Can I ask a clarifying question - is the idea that users will be expected to use `use ratatui_core`, `use ratatui_widgets` etc? or is existing code going to "just work" because we reexport everything in the `ratatui` crate? I'm assuming the latter, and that widget authors alone may want depend on `ratatui_core` alone.

**Comment by joshka** [comment]
Apps should use ratatui and will continue to work as normal. We will re-export everything in ratatui.
Widget libs should generally depend on ratatui-core to make them more resiliiant to Ratatui semver changes.

**Comment by matta** [comment]
This idea sounds good to me.

Ratatui is sometimes used for the `Terminal` abstraction alone. For example, see the Helix editor (though they vendored tui-rs into their tree...). In this new scheme it might make sense to think of the combination of ratatui-core and ratatui-{some-terminal}, without ratatui-widgets, as a use case.

**Comment by orhun** [comment]
Thanks for the example @matta, it's now more clear to me and splitting things make sense :)

> I'm 60/40 for using 0.1.0 rather than 0.29.0 for these, but happy to be convinced of the alternative approach.

Using `0.29.0` would be better in terms of knowing when the modularization occurred... I can't think of any other benefits.

**Comment by joshka** [comment]
Yeah, it would make it clear that ratatui-core 0.29.0 is the latest version of the core because it's the same as ratatui 0.29.0, but assuming that we then release 0.30.0, 0.31.0, ... without bumping core then it may get a bit confusing.

**Comment by kdheepak** [comment]
I'm not so keen on using v0.29.0 as the starting point for the new crates. We can start them at v0.1.0 and mention in the docs that the split happened in the specific version.

I do have a mild preference for making the split happen in v0.30.0 instead of v0.29.0 though, for purely psychological purposes.

**Comment by joshka** [comment]
Agreed. I'd also like to get 0.29 over the line too with whatever we put in there and then do this for 0.30.

**Comment by orhun** [comment]
Yup, it sounds good to me.

**Comment by MarcoIeni** [comment]
Hi everyone, I was chatting with Orhun about this.
This should be the only thing missing in release plz to be able to maintain the current changelog template: https://github.com/MarcoIeni/release-plz/issues/1767

It should be very easy to implement if anybody wants to give it a try 👍

**Comment by orhun** [comment]
Thanks Marco! That seems already implemented so we can probably give it a try 👍🏼

**Comment by orhun** [comment]
The RFC is accepted!

Do we have a plan before moving on with #1459, #1460 and so on?  They will conflict with the _world_ so maybe it would be nice to handle the low-hanging PRs first:

- #1442
- #1452
- #1453

**Comment by joshka** [comment]
It will be easier to rebase PRs on those than to rebase the modularization PRs on smaller PRs. In particular the move into folder PR is really small when reviewing it because it's mostly just a move. Once that's merged, 1460 gets much smaller.

**Comment by orhun** [comment]
Sounds good.

Just as an aside, for the PRs that are moving around many files, we can probably use `.git-blame-ignore-revs` for retaining the blame history: https://www.stefanjudis.com/today-i-learned/how-to-exclude-commits-from-git-blame/ (thanks to Marco for pointing that out in `git-cliff` - I still need to review/merge your PR!)

**Comment by orhun** [comment]
Let's catch up after #1460 to talk about the next steps @joshka

**Comment by joshka** [comment]
Something to consider is whether to include unstable things in ratatui-core (e.g. WidgetRef or move them back into Ratatui). Noting this quickly while I'm thinking about it to come back later to fill out details.

**Comment by joshka** [comment]
The main rationale for WidgetRef being outside of core is that we're making -core for stability. I think there's a reasonable argument that it belongs in the main crate because of that - if it is indeed possible.

**Comment by joshka** [comment]
What do we think about in ratatui-widgets putting all the widgets in proper modules rather than exporting at the root?
E.g. `ratatui_widgets::table::Row` instead of `ratatui::widgets::Row` (leaving the original location there).

**Comment by orhun** [comment]
Yup, I like it. I guess we are free to make structural adjustments in individual modules as long as we don't break the compatibility on `ratatui`.

**Comment by orhun** [comment]
Maybe we should move the modularization work from the `main` branch to another one not to block the progress on the other things. It seems like there are still some things to figure out (#1474)

**Comment by joshka** [comment]
I think I'd prefer to just acknowledge other things are blocked a bit for now, and take the time to get it right. There's nothing critical to release soon afaik (correct me if I'm wrong there).

**Comment by orhun** [comment]
Yeah, that sounds good. I was panicking there for a moment.

Asking in advance, can you invite @ratatui/maintainers to the following crates?

- https://crates.io/crates/ratatui-widgets
- https://crates.io/crates/ratatui-core

**Comment by joshka** [comment]
> Asking in advance, can you invite [@ratatui/maintainers](https://github.com/orgs/ratatui/teams/maintainers) to the following crates?

Done as well as the backend crates

**Comment by joshka** [comment]
Regarding Frame, in tui-prompts I have https://github.com/joshka/tui-widgets/blob/88e3bdd10d29ec6c243918ceb5f9380770106e61/tui-prompts/src/prompt.rs#L12-L21

```rust
pub trait Prompt: StatefulWidget {
    /// Draws the prompt widget.
    ///
    /// This is in addition to the [`StatefulWidget`] trait implementation as we need the [`Frame`]
    /// to set the cursor position.
    ///
    /// [`StatefulWidget`]: ratatui::widgets::StatefulWidget
    /// [`Frame`]: ratatui::Frame
    fn draw(self, frame: &mut Frame, area: Rect, state: &mut Self::State);
}
```

The rationale of this was to be able to let the prompts directly adjust the frame cursor position.
I think there might be other places where access to the frame from non-app libs might be useful. I don't know of specific use cases for this and haven't really thought much about the downside risk (is Frame an unstable type, likely not, but we've had some discussions over things which could be added to Frame recently, so this risk is low but not zero).

I think this says Frame belongs in -core.

**Comment by joshka** [comment]
I manually released -alpha.0 versions of each of the crates earlier. Doing it manually (as opposed to using release automation) highlighted a few things that weren't obvious, and were easy-ish to fix. https://github.com/ratatui/ratatui/pull/1525 tracks the necessary changes I made to get this publishable.

https://github.com/ratatui/ratatui-macros/pull/75 updates and deploys ratatui-macros (should we bring this crate into this monorepo perhaps? leaning towards probably not)

https://github.com/joshka/tui-widgets/pull/47 is a wip attempt to update tui-widgets to use the -alpha releases. tui-prompts is blocked on Frame, and tui-popup is TODO in a separate PR.

**Comment by orhun** [comment]
Thanks for the alpha releases and testing! Maybe we should ping some fellow library authors soon to get some feedback on this as well.

> Doing it manually (as opposed to using release automation)

I can focus on the release automation next week probably. These days I'm still a bit stuck with refactoring the examples & other stuff sadly.

> I think this says Frame belongs in -core.

I'm okay with moving `Frame` to core although I cannot tell if there are any side effects to that.

Just to make sure, is this only needed because your custom trait implementation needs to re-use the `Frame` to set the cursor position?

I'm also wondering if we need to move anything else, it would probably help if we do a search on GitHub to see if anyone is using any other types/traits in a custom way.

**Comment by thscharler** [comment]
StatefulWidgetRef is not in core yet? Is this going to be abandoned or has it just been missed?

**Comment by joshka** [comment]
I'm going to use the term terms WidgetRef for this response to refer to both WidgetRef and StatefulWidgetRef.

For now, WidgetRef remains in the main ratatui crate and not in ratatui-core. This was an intentional choice, but certainly one that we can discuss if it's the wrong one. We didn't want these in in ratatui-core as we don't want to have to release multiple versions of core to update it as we work through choosing the right abstractions. We want ratatui-core to be a stable place that makes it easier for widget libraries to generally not have to be updated any time we make a change to the built-in ratatui widgets etc.

You should also note that for 0.30, we also reversed the way the the blanket implementation of WidgetRef was implemented. Instead of making a blanket `impl of Widget for &W where W: impl WidgetRef` we reversed this to a blanket `impl WidgetRef for W where &W: Widget`. Doing this meant that we were able to remove the need to implement WidgetRef from all the Widgets in ratatui-widgets, while still making it possible to use these widgets in context where they need to be boxed.

I'd suggest that any widget libraries that need boxed widgets / widgetrefs in general should test out the alpha release and check that these abstractions work for their use case.

The main reason that these parts aren't yet stabilized yet is that we don't really have a strong opinion on what abstractions make sense to lock into the way widgets work generally. Both naming and exact behavior are key parts of this. For the most part we've taken the tack of supporting just enough to make it work, rather than investing the time in to design a full solution to solve all the problems of widgets more generally. https://github.com/ratatui/ratatui/issues/1287 is the tracking issue right now, but we don't really have a good discussion issue for this right now - maybe https://github.com/ratatui/ratatui/issues/996? At some point we should circle back around to solve it.

**Comment by thscharler** [comment]
All is well, I get the main point that you don't want unstable things in core. Fine with me. 

In all my code I never saw a strong need for those two traits, I just included them to be on the safe side. 

Apart from that I could build rat-salsa with only using ratataui-core, ratatui-widgets and ratatui-crossterm included. 

And ratatui-widgets is mostly needed for Block. Is Block not considered a core component? I would have thought so as ubiquitous as it is?

**Comment by joshka** [comment]
> In all my code I never saw a strong need for those two traits, I just included them to be on the safe side.

The main (only?) need for the -Ref traits is where Widget/StatefulWidget don't work. The only place I know so that cannot be implemented using the standard Widget trait is when you some sort of container that accepts dyn Widget. You can't call the render method on `Box<dyn Widget>`. E.g. you can't write:

```rust
let widgets: Vec<Box<dyn Widget>> = vec![Box::new("hello"), Box::new("world")];
for (area, widget) in zip(areas, widgets) {
    widget.render(area, &mut buf);
}
```

But you can with WidgetRef. So unless you've got that pattern going on in rat-salsa, I'd advise `impl Widget for &YourWidgetType` as a general rule.

> And ratatui-widgets is mostly needed for Block. Is Block not considered a core component? I would have thought so as ubiquitous as it is?

I wanted to avoid putting block in -core, mainly because we are actively changing things there still. Things to do with how borders work are a particular thing that I know will probably change. It's used a lot - that's true. Perhaps at some point in fhe future we should pull each widget out into its own library (like what I've been doing with tui-widgets). This would be a lot of overhead for now, but the key ones that really could do with this would be table and list. Maybe some time later in the year we should come back to that idea though.

**Comment by thscharler** [comment]
Ah yes, the container of Widget problem. 
I only have that in the rat-dialog crate and use my own traits for that one. 

I will heed your advice for the regular Widgets and add the impl for &Widgets.

**Comment by joshka** [comment]
@thscharler As an aside - I wonder if you'd consider writing up something about rat-salsa to put in https://ratatui.rs/concepts/application-patterns/ ? I recall taking a look at the approaches you took for events a while back and I didn't like some of the design choices (purely from an idealistic perspective, but I don't recall the exact details right now), but from a pragmatic perspective though, the stuff that you're building looks great and should be something that people consider when building good looking Ratatui apps.

**Comment by thscharler** [comment]
I have written a mdbook thing for rat-salsa, but it's outdated by now and I'm going to rewrite from scratch. And I'm now starting to go over all the API and fill in the gaps in the docs. 

There will be some design chapter, I think. 

There should be something in the next weeks, I'll come back to you.

**Comment by joshka** [comment]
Closing this out as completed by many PRs. We can open it back up if there's specific discussion on any of the points above, but I'd be inclined to open new issues / discussions if we run into problems during beta testing the modularized ratatui.
